Mathlib/LinearAlgebra/Dimension.lean

```diff
@@ -547,6 +547,156 @@ theorem rank_pos_iff_nontrivial : 0 < Module.rank R M ↔ Nontrivial M :=
 
 end RankZero
 
+section SurjectiveInjective
+
+section Module
+
+variable {R : Type w} {M : Type v} [Ring R] [AddCommGroup M] [Module R M]
+  {R' : Type w'} {M' : Type v'} [Ring R'] [AddCommGroup M'] [Module R' M']
+
+/-- If `M / R` and `M' / R'` are modules, `i : R' → R` is a map which sends non-zero elements to
+non-zero elements, `j : M →+ M'` is an injective group homomorphism, such that the scalar
+multiplications on `M` and `M'` are compatible, then the rank of `M / R` is smaller than or equal to
+the rank of `M' / R'`. As a special case, taking `R = R'` it is
+`LinearMap.lift_rank_le_of_injective`. -/
+theorem lift_rank_le_of_injective_injective (i : R' → R) (j : M →+ M')
+    (hi : ∀ s, i s = 0 → s = 0) (hj : Injective j)
+    (hc : ∀ (s : R') (x : M), j (i s • x) = s • j x) :
+    lift.{v'} (Module.rank R M) ≤ lift.{v} (Module.rank R' M') := by
+  cases subsingleton_or_nontrivial R'
+  · nontriviality R
+    haveI := Module.subsingleton R' M'
+    haveI := hj.subsingleton
+    have : Module.rank R M = 0 := rank_eq_zero_iff.2 fun _ ↦ ⟨1, one_ne_zero, Subsingleton.elim _ _⟩
+    simp only [rank_subsingleton, this, lift_zero, lift_one, zero_le]
+  rw [Module.rank_def R M]
+  refine lift_iSup_le (bddAbove_range.{v, v} _) fun S ↦ ?_
+  obtain ⟨_, h⟩ := S
+  exact cardinal_lift_le_rank_of_linearIndependent (h.map_of_injective_injective i j hi
+    (fun x ↦ by rw [← _root_.map_zero (f := j)]; exact hj (a₁ := x) (a₂ := 0)) hc)
```

Comment:
You can avoid nontrivial/subsingleton case splits by following the proof of `lift_rank_le_of_injective`:
```suggestion
  simp_rw [Module.rank, lift_iSup (bddAbove_range.{v', v'} _), lift_iSup (bddAbove_range.{v, v} _)]
  exact ciSup_mono' (bddAbove_range.{v', v} _) fun ⟨s, li⟩ ↦ ⟨⟨j '' s,
    (li.map_of_injective_injective i j hi (fun x hx ↦ hj <| by rwa [_root_.map_zero]) hc).image⟩,
      lift_mk_le'.mpr ⟨(Equiv.Set.image j s hj).toEmbedding⟩⟩
```